### PR TITLE
Fix #140: factory default init_grace_tokens for lastrec (attention sinks)

### LINKED ISSUE #140: Evicting KV cache corrupts generation (garbage output once prompt > cache_length)
**Describe the bug**

Autoregressive generation through an **evicting** KV cache (e.g. `lastrec`, `h2o`,
and quantized variants) produces garbage — the model emits EOS/newlines
immediately — whenever the prompt is longer than `cache_length` (so the prefill
spans multiple chunks and eviction occurs). With a cache large enough to hold the
whole prompt (no eviction), generation is correct and matches a plain dense /
no-cache forward. This is a **cliff**, not gradual lossy degradation, and it fires
even when only *irrelevant* tokens are evicted, so it is a mechanism bug rather
than expected sparse-attention information loss.

Since eviction is the whole point of these caches for long context (and eval
normally uses `cache_length < sequence_length`), this affects long-context
generation/eval broadly.

**Evidence (Qwen2.5-0.5B-Instruct, A10G, bf16, eager SDPA)**

Controlled key-value retrieval where the answer is the **last** record (so
`lastrec`, which keeps the most-recent tokens, provably retains the needle *and*
the question), ~2062-token prompts, accuracy = substring match of the value:

| cache_length | eviction? | accuracy |
|---:|:--:|---:|
| 2112 | no  | **0.90** |
| 1900 | yes (drops only first ~162 filler tokens) | **0.00** |
| 1536 | yes | 0.00 |
| 1024 | yes | 0.00 |
| 512  | yes | 0.00 |

- The collapse happens at the *first* eviction, dropping only early filler tokens
  that are irrelevant to a last-record query — inconsistent with legitimate
  lossiness.
- Both `lastrec-default` and `lastrec-torch-quantized8` collapse to 0.00, so it is
  **not** a quantization issue.
- `h2o-torch-quantized8` behaves identically.

**Localization**

Only the first prefill chunk (`input_pos == 0`) takes the `_prefill` path
(attends to itself). Every subsequent chunk — which is where eviction happens —
goes through the `_forward` path and uses `token_positions()` to build the
attention mask (`build_mask_slice` in `keys_values/attention/base.py`). The bug
requires a multi-chunk prefill, so it is in that shared masked-attention /
eviction path. `token_positions()` itself appears to return correct original
positions, which points at the mask-slice / eviction-slot bookkeeping. The
first *generated* token (straight from the prefill final-position logits) is
already wrong, so the corruption is in the prefill, before any decode step.

**Severity is scale-dependent (why it may slip past tests):** a tiny 2-layer
random model in fp32 shows only a mild final-logit perturbation under eviction
(max|Δ| ≈ 0.08, argmax preserved), whereas the real 24-layer model in bf16
collapses to garbage. Unit tests on small models may therefore pass.

**To reproduce**

```python
import torch, lightning as L
from litgpt.prompts import PromptStyle, has_prompt_style, load_prompt_style
from litgpt.tokenizer import Tokenizer
from litgpt.utils import auto_download_checkpoint, load_checkpoint
from keys_values.config import Config
from keys_values.data.constants import LIT_MODEL_FNAME
from keys_values.kvcache.factory import KVCacheFactory, deallocate_kv_cache_buffers_of_model
from keys_values.long_context import LongContextInferenceModel
from keys_values.model import GPT
from keys_values.generate.base import batched_generate_fn
from keys_values.utils import VerbosityLevels

fab = L.Fabric(devices=1, accelerator="cuda", precision="bf16-true")
cd = auto_download_checkpoint(model_name="Qwen/Qwen2.5-0.5B-Instruct")
tok = Tokenizer(cd); cfg = Config.from_file(cd / "model_config.yaml")
ps = load_prompt_style(cd) if has_prompt_style(cd) else PromptStyle.from_config(cfg)
pad = tok.processor.token_to_id("<|endoftext|>") or int(tok.eos_id)
with fab.init_module(empty_init=True): m = GPT(cfg)
load_checkpoint(fab, m, cd / LIT_MODEL_FNAME); m.to(fab.device)

# ~2000-token prompt of "key is value" records; the ANSWER is the LAST record.
import random; rng = random.Random(0)
recs = [f"The value for key '{''.join(rng.choice('abcdefghij') for _ in range(6))}' is {rng.randint(100000,999999)}." for _ in range(180)]
answer = "424242"; recs.append(f"The value for key 'zzzzzz' is {answer}.")
body = ("Records:\n" + " ".join(recs) +
        "\n\nQuestion: what is the value for key 'zzzzzz'? Reply with only the value.\nAnswer:")
ids = tok.encode(ps.apply(body), device=fab.device).unsqueeze(0)
L_ = ids.shape[1]

def gen(cache_len):
    deallocate_kv_cache_buffers_of_model(m)
    m.assign_kv_caches(KVCacheFactory.create(gpt_model=m, name="lastrec-default",
        max_batch_size=1, cache_length=cache_len, dtype=torch.bfloat16))
    m.eval(); m.max_seq_length = L_ + 16
    inf = LongContextInferenceModel(m, head_model=None, chunk_size=min(512, cache_len),
                                    verbose=VerbosityLevels.NONE)
    out = list(batched_generate_fn(model=inf, prompts=ids, max_returned_tokens=8,
        sample_args=dict(temperature=1.0, top_k=1, top_p=1.0), stop_tokens=(), ignore_index=pad))
    return tok.decode(torch.cat(out, dim=1)[0])

print("prompt_len", L_, "answer", answer)
print("no-evict (cl=L+32):", repr(gen(L_ + 32)))   # -> '424242'
print("evict   (cl=1024) :", repr(gen(1024)))       # -> garbage (e.g. '\n\n\n...')
```

Expected: both should return `424242` (the last record is retained under
`lastrec` either way). Actual: the no-eviction case is correct; the eviction case
returns garbage.

**Environment**: Ubuntu 24.04, A10G, torch 2.13.0+cu130, `lastrec` / `h2o` caches,
eager SDPA (FlashInfer disabled to rule the kernel out). Reproduced on the
`grpo-cache-reuse` branch of `mmjerge/keys_values`.

**Related**: possibly the same underlying cause as #99 (`smart-lastrec`) and #5
(`h2o`/`qh2o` scoring); distinct from #139 (`ln_f`).

### PR DESCRIPTION
**Pull request by mmjerge**
Split out of #142 per review, and trimmed to the conservative scope discussed there: **only `lastrec` gets a factory default** (`init_grace_tokens = min(16, cache_length // 8)`); the H2O family is unchanged.

## Why

`lastrec` evicts the *initial* tokens first. Trained LLMs use those tokens (BOS, template header) as attention sinks, and evicting them collapses generation to immediate EOS/garbage — every evicting configuration scored 0.000 on HELMET before this fix (measured data in #140). Pinning a few initial tokens restores generation. H2O-type caches are not affected: `keep_initial_fraction` plus accumulated attention scores already keep the sink tokens resident.

## Scope

- `keys_values/kvcache/factory.py`: `_apply_safe_eviction_defaults` applied in the three factory entry points; callers can pass an explicit value (e.g. `0`) to override.
- `keys_values/kvcache/test_utils.py`: unit tests exercise raw eviction mechanics, so `create_kv_cache` opts out (`init_grace_tokens=0`) unless a test asks otherwise.

## Testing

`test/kvcache`: 956 passed, 85 skipped.

#142 no longer touches these files and is now RL-only.

### CODE REVIEW

File: keys_values/kvcache/factory.py [hunk 1]
```diff
@@ -65,6 +65,43 @@
     if do_def or quant != "default"
 }
 
+# Safe eviction default for `lastrec`, applied by the factory if the caller
```
Comment (mseeger) [inline keys_values/kvcache/factory.py:68]: I'd prefer to have the default being changed in the `LastRecentlyInsertedKVCache` class.

Let me just do this, and close this PR (but thanks for figuring this out!).

**Comment by mseeger** [comment]
I am closing this and will make the change directly. Thanks.
